[Chen Yuan] iP

cyyeu opened the pull request:
# Duke
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ *SUPER FAST* to use

All you need to do is,
1. download it from [here](https://github.com/cyyeu/ip/releases/download/A-Jar/Duke.jar).
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing todos
- [x] Managing deadlines 
- [x] Managing events 
- [x] GUI
##
If you Java programmer, you can use it to practice Java too. Here's the main method:
```java
public static void main(String[] args) {
        String filePath = Storage.DEFAULT_FILE_PATH;
        if (args.length >= 1) {
            filePath = args[0];
        }
        new Duke(filePath).run();
    }
```

Commit:
Add Gradle support

cyyeu committed:
Level-1

cyyeu committed:
Level-2

cyyeu committed:
Level-3

cyyeu committed:
Level-4

cyyeu committed:
Add tests

cyyeu committed:
Level-5

cyyeu committed:
Level-6

cyyeu committed:
Implement Save feature

cyyeu committed:
Refractor event and deadline parser

cyyeu committed:
Refractor Deadline class to use java.time.LocalDate

cyyeu committed:
Merge branch 'branch-Level-7'

cyyeu committed:
Merge branch 'branch-Level-8'

cyyeu committed:
Refractor code structure to adhere to OOP and java packages conventions

cyyeu committed:
Add JUnit Tests

cyyeu committed:
Add manifest

cyyeu committed:
Add javadocs to Parser.java

cyyeu committed:
Fix java styling conventions

cyyeu committed:
Implement find feature

cyyeu committed:
Merge branch 'branch-A-JavaDoc'

cyyeu committed:
Merge branch "branch-A-CodingStandard"

cyyeu committed:
Merge branch "branch-Level-9"

cyyeu committed:
Refractor wildcard imports to use explict import statements

cyyeu committed:
Add comments

cyyeu committed:
Trim user input

cyyeu committed:
Merge remote-tracking branch 'origin/add-gradle-support' into feat/gradle

cyyeu committed:
Set up build.gradle

cyyeu committed:
Merge branch 'feat/gradle'

cyyeu committed:
Set up checkstyle config

cyyeu committed:
Fix all styling errors according to Checkstyle

dt-td commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,64 @@
+        new Duke(filePath).run();
+    }
+
+    /**
+     * Solution adapted with minor modifications at 
+     * https://nus-cs2103-ay2122s1.github.io/website/schedule/week3/project.html#a-moreoop
+     */
+    private void run() {
```
I like how you included the source in your javadoc 👍

alissayarmantho replied:
+1

dt-td commented on src/main/java/duke/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,43 @@
+        this.taskId = taskId;
+    }
+
+    /**
+     * Executes the delete command by deleting the task in the in-memory list,
+     * and deleting the corresponding task on the txt file.
+     *
+     * @param tasks The current list of tasks.
+     * @param ui The UI of the Duke app.
+     * @param storage The storage manager for the Duke app.
+     * @throws InvalidTaskSelectedException Thrown exception when user inputs invalid task Id.
+     * @throws IOException Thrown IO exception when trying to delete tasks on storage.
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage)
```
I like your javadoc descriptions, they are very clear and I can understand them easily the first time I read it 👍

dt-td commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,187 @@
+package duke.parser;
+
+import duke.commands.AddTaskCommand;
+import duke.commands.Command;
+import duke.commands.DeleteCommand;
+import duke.commands.DoneCommand;
+import duke.commands.ExitCommand;
+import duke.commands.FindCommand;
+import duke.commands.InvalidArgumentsException;
+import duke.commands.ListCommand;
+import duke.commands.UnknownCommandException;
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Todo;
+import java.time.LocalDate;
+import java.time.format.DateTimeParseException;
```
I like how you ordered the import statements. They are consistent. At one glance I can understand the code base easily. 👍

dt-td commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,187 @@
+     * @return The corresponding command.
+     * @throws UnknownCommandException Thrown if the command keyword has no match.
+     * @throws InvalidArgumentsException Thrown if there are invalid arguments.
+     * @throws UnableToParseException Thrown if there was an error parsing the command.
+     */
+    public static Command parseCommand(String input)
+            throws UnknownCommandException, InvalidArgumentsException, UnableToParseException {
+        // index 0 has command, index 1 has command arguments (if applicable)
+        String[] splitInput = input.trim().split(" ", 2);
+        String stringCmd = splitInput[0];
+        switch (stringCmd) {
+        case "todo":
```
I like how you placed a line break here to improve readability. 👍

dt-td commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,97 @@
+            }
+        }
+
+        return matches;
+    }
+    
+
+    private void validateTaskId(int id) throws InvalidTaskSelectedException {
+        int index = id - 1;
+        if (index < 0 || index >= this.tasks.size()) {
```
Perhaps a single line break here?

dt-td reviewed (commented):
LGTM! Just some nits to fix. 👍

alissayarmantho commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,64 @@
+            } catch (DukeException e) {
+                ui.printErr(e.getMessage());
+            } catch (Exception e) {
+                ui.printErr("Unexpected Error: " + e.getMessage());
```
Interesting to catch all kinds of exceptions...Usually, we don't do Pokemon Catch All Exception, but I guess in this context it is fine.

alissayarmantho commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,187 @@
+import duke.commands.UnknownCommandException;
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Todo;
```
`
`
Add new line between duke related imports and java imports

alissayarmantho commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,159 @@
+    /**
+     * Default file path where the storage is saved.
+     */
+    public static final String DEFAULT_FILE_PATH = "data/duke.txt";
```
You might want to use a relative path to the current directory of the program

alissayarmantho commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,159 @@
+     * @throws IOException unexpected IO error.
+     */
+    public void update(int index, Task task) throws IOException {
+        List<String> fileContent = new ArrayList<>(Files.readAllLines(this.path));
```
Actually, this seems a bit wasteful. Maybe you can just update/delete when the program exits, so you don't need to keep reading all lines and writing the entire task lists on a simple update/delete

cyyeu replied:
Hi, I think this has to been discussed in the forum [here](https://github.com/nus-cs2103-AY2122S1/forum/issues/62). The downside to that is that "The program might crash or be stopped by the user then the tasks wouldn't be saved"

alissayarmantho replied:
You can add your store to hard disk function to your pokemon catch all try-catch statement and save the file...

Or you can add a shutdown hook
```Runtime.getRuntime().addShutdownHook(new Thread(() -> storeExistingTaskList()));```

alissayarmantho commented on src/test/java/duke/task/TaskTest.java:
```diff
@@ -0,0 +1,29 @@
+package duke.task;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Nice to know you are actually testing something important :)

alissayarmantho reviewed (changes requested):
Overall, quite good. Your code follows the coding standard of this module.

cyyeu committed:
Implement GUI feature

cyyeu committed:
Merge branch 'branch-Level-10'

jyrw commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,54 @@
+
+    Duke(Stage stage) {
+        try {
+            this.ui = new Ui(stage, this);
+            ui.print("Initializing Duke...");
+            this.storage = new Storage();
+            this.tasks = this.storage.load();
```
Very very minor, but I think the use of the ```this``` keyword can be avoided here, as per [this](https://se-education.org/guides/conventions/java/index.html#variables).

jyrw reviewed (commented):
Very clean code! I really can't find any code quality violations.

cyyeu committed:
Add assertions in Duke

Assertions were not added in Duke.

Added various assertions to ensure invariants are checked (objects/strings not null)

Allow the runtime to verify these assertions so as to catch bugs early

cyyeu committed:
Refractor: Improve code quality

Previously, there were multiple variables that were hardcoded inside the functions. Also, the Storage::load function had to parse the strings read from the file.

The said variables were extracted out and made static final where applicable.
The parsing of strings from storage is now extracted out into a new package at duke.parser.storage.StorageParser

This is done to improve code readability, and to fully extract out all parsing work to the parser package. This allows for the code in the Storage::load function to be much cleaner and shorter.

cyyeu committed:
Merge pull request #1 from cyyeu/branch-A-Assertions

Add assertions in Duke

cyyeu committed:
Merge branch 'master' into branch-A-CodeQuality

cyyeu committed:
Merge pull request #2 from cyyeu/branch-A-CodeQuality

Refractor: Improve code quality

cyyeu committed:
Set up Github Actions CI

cyyeu committed:
Feature: Update command

cyyeu committed:
Merge branch 'feat/update-command'

cyyeu committed:
Fix typo in comment

cyyeu committed:
Fix styling

cyyeu committed:
Improve GUI layout

cyyeu committed:
Improve JavaDocs phrasing

cyyeu committed:
Update user guide README

cyyeu committed:
Set theme jekyll-theme-slate

cyyeu committed:
Improve user guide README

cyyeu committed:
Rename UI.png to Ui.png

cyyeu committed:
Fix numbering in user guide

cyyeu committed:
Update user guide

cyyeu committed:
Fix user guide for github pages

cyyeu committed:
Fix user guide for github pages

cyyeu committed:
Fix user guide for github pages

cyyeu committed:
Refractor updating deadline format in user guide

cyyeu committed:
Fix user guide block quote indentation

cyyeu committed:
Fix typo in user guide

cyyeu committed:
Set theme jekyll-theme-hacker

cyyeu committed:
Update user guide

cyyeu committed:
Update user guide
